src/module_library/daylength_calculator.h

```diff
@@ -0,0 +1,117 @@
+ * sets the solar elevation (degrees) at which the sun rises and sets. `p = 0`
+ * means the horizon.
+ *
+ * Table 1: Daylength definitions defined by the position of the sun with respect to the horizon
+ * | | Daylength definition |  P |
+ * | :--- |:--- | ---: |
+ * | (with and without twilight) | (degrees) |
+ * | 1 | Sunrise/Sunset (center at horizon) |  0.0 |
+ * | 2 | Sunrise/Sunset (top/bottom at horizon)  | 0.26667 |
+ * | 3 | Sunrise/Sunset (apparent top/bottom, US government definition)  | 0.8333 (a) |
+ * | 4 | With civil twilight | 6.0 |
+ * | 5 | With nautical twilight | 12.0 |
+ * | 6 | With astronomical twilight | 18.0 |
+ * (a) This value is the summation of the radius of the sun (in
+ * degrees as seen from Earth) plus the adopted value for the
+ * refraction of the light through the atmosphere of 34 minutes
+ * (Astronomical Almanac 1992)
+ *
+ * Reference
+ *
+ * - William C. Forsythe, Edward J. Rykiel, Randal S. Stahl, Hsin-i Wu, Robert M.
+ *   Schoolfield. "A model comparison for daylength as a function of latitude and
+ *   day of year." Ecological Modelling, Volume 80, Issue 1, 1995, Pages 87-95,
+ *   https://doi.org/10.1016/0304-3800(94)00034-F.
+ */
+class daylength_calculator : public direct_module
+{
+   public:
+    daylength_calculator(
+        state_map const& input_quantities,
+        state_map* output_quantities)
+        : direct_module{},
+
+          // Get references to input quantities
+          lat{get_input(input_quantities, "lat")},
+          solar_dec{get_input(input_quantities, "solar_dec")},
+          p{get_input(input_quantities, "sunrise_sunset_elevation")},
+          // Get pointers to output quantities
+          daylength{get_op(output_quantities, "day_length")}
+
+    {
+    }
+    static string_vector get_inputs();
+    static string_vector get_outputs();
+    static std::string get_name() { return "daylength_calculator"; }
+
+   private:
+    // References to input quantities
+    const double& lat;
+    const double& solar_dec;
+    const double& p;
+
+    // Pointers to output quantities
+    double * daylength;
+    // Main operation
+    void do_operation() const;
+};
+
+string_vector daylength_calculator::get_inputs()
+{
+    return {
+        "lat",      // degrees
+        "solar_dec", // degrees
+        "sunrise_sunset_elevation" // degrees
+    };
+}
+
+string_vector daylength_calculator::get_outputs()
+{
+    return {
+        "day_length" // hr
+       };
+}
+
+void daylength_calculator::do_operation() const
+{
+    constexpr double pi = 3.14159265359;
+    constexpr double to_rad = pi/180;
+    double offset = std::sin(p * to_rad);
+    double lam = lat * to_rad;
+    double phi = solar_dec  * to_rad;
+    double num  = offset  + std::sin(lam) * std::sin(phi);
+    double denom = std::cos(lam) * std::cos(phi);
+    double u = std::max(std::min(num/denom, 1.0), -1.0);
+    double D = 24 - 24/pi * std::acos(u);
```

Comment:
Please add comments showing the units of these calculated variables. It seems that `D` is in hours, `lam` is in radians, `denom` is dimensionless, etc.